Fix #4949 and Avoid Errno::EBUSY on docker containers

andrewjamesbrown opened the pull request:
This commit fixes https://github.com/chef/chef/issues/4949 by adding
ohai-based docker detection code and statically checking for
/etc/hostname or /etc/resolv.conf.

cc @lamont-granquist

lamont-granquist commented on lib/chef/provider/file.rb:
```diff
@@ -193,6 +199,12 @@ def manage_symlink_access?
       private
 
+      # Return true if we are running in a docker guest, or false otherwise.
+      def docker_guest?(node)
```
ideally i'd like to see `#docker?` pulled into Chef::DSL::Universal from chef-sugar:  

https://github.com/sethvargo/chef-sugar/blob/b9b3b643c27a0828f56d7355bb4d05577bd90a38/lib/chef/sugar/docker.rb

lamont-granquist replied:
(i don't know exactly what to do about back-compat with chef-sugar and how to deal with the way that chef-sugar has that method which takes a node as an argument or what that use case covers...)

lamont-granquist commented on lib/chef/provider/file.rb:
```diff
@@ -67,7 +67,13 @@ class File < Chef::Provider
+                                                 new_resource.path == "/etc/hostname")
+            @deployment_strategy = Chef::FileContentManagement::Deploy.strategy(false)
+          else
+            @deployment_strategy = Chef::FileContentManagement::Deploy.strategy(new_resource.atomic_update)
```
this could probably look more like:

```ruby
atomic_updates = special_docker_files? ? false : new_resource.atomic_updates
@deployment_strategy = Chef::FileContentManagement::Deploy.strategy(atomic_updates)
```

also should probably compare against `Pathname.cleanpath(new_resource.path)` and/or use `File.expand_path()`

coderanger replied:
I feel like an explicit value in `new_resource.atomic_updates` should always override, just that the default becomes path-dependent.

lamont-granquist replied:
yeah, probably should.

andrewjamesbrown replied:
Added this behaviour with d801443.

andrewjamesbrown commented:
I've updated the implementation so that `new_resource.atomic_updates` always overrides the default, and also moved the implementation of `docker?` into `lib/chef/dsl/platform_introspection.rb`.  I wasn't able to use the same code as chef-sugar, as Travis was complaining bitterly about the `File.exists?` and mocking that everywhere with docker-specific code didn't seem to be the best option, so I'm just looking at Ohai data.

lamont-granquist commented:
mmm yeah, file provider unit tests and mocking `File.exists?`...   there be dragons...

lamont-granquist commented:
👍

smurawski commented:
👍
